packages/Chem/src/analysis/molecular-matched-pairs/mmp-cliffs.ts

```diff
@@ -0,0 +1,148 @@
+import {BitArrayMetrics, BitArrayMetricsNames} from '@datagrok-libraries/ml/src/typed-metrics';
+import {chemSpace} from '../chem-space';
+import {debounceTime} from 'rxjs/operators';
+import {getInverseSubstructuresAndAlign} from './mmp-mol-rendering';
+import {MMP_COLNAME_FROM, MMP_COLNAME_TO, MMP_COL_PAIRNUM_FROM, MMP_COL_PAIRNUM_TO, MMP_COLNAME_CHEMSPACE_X,
+  MMP_COLNAME_CHEMSPACE_Y, MMP_STRUCT_DIFF_FROM_NAME, MMP_STRUCT_DIFF_TO_NAME} from './mmp-constants';
+
+export function getMmpScatterPlot(table: DG.DataFrame, activities: DG.ColumnList, maxActs: number[]) :
+[sp: DG.Viewer, sliderInputs: DG.InputBase[], sliderInputValueDivs: HTMLDivElement[], colorInputs: DG.InputBase[]] {
+  const colX = DG.Column.float('~X', table.rowCount);
+  const colY = DG.Column.float('~Y', table.rowCount);
+  table.columns.add(colX);
+  table.columns.add(colY);
+  const sp = DG.Viewer.scatterPlot(table, {
+    x: '~X',
+    y: '~Y',
+    zoomAndFilter: 'no action',
+    //color: activities.name,
+    showXSelector: false,
+    showYSelector: false,
+    markerDefaultSize: 7,
+  });
+
+  const sliderInputs = new Array<DG.InputBase>(maxActs.length);
+  const sliderInputValueDivs = new Array<HTMLDivElement>(maxActs.length);
+  const colorInputs = new Array<DG.InputBase>(maxActs.length);
+
+  for (let i = 0; i < maxActs.length; i ++) {
+    const sliderInput = ui.sliderInput(activities.byIndex(i).name, 0, 0, maxActs[i]);
+    sliderInput.root.style.marginLeft = '6px';
+    const sliderInputValueDiv = ui.divText(sliderInput.stringValue, 'ui-input-description');
+    sliderInput.addOptions(sliderInputValueDiv);
+    sliderInputs[i] = sliderInput;
+    sliderInputValueDivs[i] = sliderInputValueDiv;
+    const colorInput = ui.colorInput('', '#FF0000');
+    colorInput.root.style.marginLeft = '6px';
+    colorInputs[i] = colorInput;
+  }
+
+  return [sp, sliderInputs, sliderInputValueDivs, colorInputs];
+}
+
+function drawMolPair(molecules: string[], indexes: number[], substruct: (ISubstruct | null)[], div: HTMLDivElement,
+  parentTable: DG.DataFrame, tooltip?: boolean) {
+  ui.empty(div);
+  const hosts = ui.divH([]);
+  if (!tooltip)
+    hosts.classList.add('chem-mmpa-context-pane-mol-div');
+  const canvasWidth = 170;
+  const canvasHeight = 75;
+  for (let i = 0; i < 2; i++) {
+    const imageHost = ui.canvas(canvasWidth, canvasHeight);
+    imageHost.onclick = () => {parentTable.currentRowIdx = indexes[i];};
+    drawMoleculeToCanvas(0, 0, canvasWidth, canvasHeight, imageHost, molecules[i], '', undefined, substruct[i]);
+    hosts.append(imageHost);
+  }
+  div.append(hosts);
+};
+
+export function fillPairInfo(line: number, linesIdxs: Uint32Array, activityNum: number, pairsDf: DG.DataFrame,
+  diffs: Array<Float32Array>, parentTable: DG.DataFrame,
+  rdkitModule: RDModule, propPanelViewer?: FormsViewer):
+  HTMLDivElement {
+  const div = ui.divV([], {style: {width: '100%', height: '100%'}});
```

Comment:
Prefer to use CSS where possible, it's a lot easier to adjust for someone else (like Alex)